doc/config.md
## Config Component

The configuration module is used to define dependency parameters when HTTP service starts, defined as follows:

``` go
type Config struct {
	HttpAddr         string `json:"http_addr,omitempty"` // e.g. "127.0.0.1", "unix:/tmp/app.sock"
	HttpPort         uint16 `json:"http_port,omitempty"` // e.g. 8080
	HttpTimeout      uint16 `json:"http_timeout,omitempty"`
	UrlBasePath      string `json:"url_base_path,omitempty"`
	CookieKeyLocale  string `json:"cookie_key_locale,omitempty"`
	CookieKeySession string `json:"cookie_key_session,omitempty"`
}
```

Parameter Description

| Field | Type | Default | Required | Description |
|----|----|----|----|----|
| HttpAddr | string | No | Empty | Set to publish services via Unix domain socket |
| HttpPort | int | No | 8080 | Set to publish services via TCP port |
| HttpTimeout | int | No | 30 | Set HTTP connection timeout in seconds |
| UrlBasePath | string | No | / | Set root URL path for HTTP service access, default is / |
| CookieKeyLocale | string | No | lang | When i18n is enabled, httpsrv will set language package parameters in cookie with default field name `lang`. This value can customize cookie field name for saving |
| CookieKeySession | string | No | access_token | When Session is enabled, httpsrv will set user status Session value information in cookie with default field name `access_token`. This value can customize cookie field name for saving |

Config is a built-in item of [Service](service.md) and can be referenced via Service, such as:

``` go
package main

import ( 
	"github.com/hooto/httpsrv"
)

func main() {
	// Reference via global Service instance 
	httpsrv.DefaultService.Config.HttpPort = 8080

	// Reference via new Service instance
	srv := httpsrv.NewService()
	srv.Config.HttpPort = 8081
}
```

## Extended Configuration Items

On the basis of `type Config struct` data definition, some dynamic interfaces are extended to extend configuration items.

### Register New Template Built-in Function (Optional)

``` go
package main

import ( 
	"strings"
	"github.com/hooto/httpsrv"
)

func ExampleUpper(v string) string {
	return strings.ToUpper(v)
}

func main() {
	var conf httpsrv.Config
	// ...
	conf.RegisterTemplateFunc("upper", ExampleUpper)
}
```

> Note: System has built-in commonly used view template (View Template) functions by default. For details, refer to code file [template-func.go](https://github.com/hooto/httpsrv/blob/master/template-func.go)

doc/zh-CN/config.md
## Config 组件

配置模块用于定义 HTTP 服务启动时的依赖参数，定义如下:

``` go
type Config struct {
	HttpAddr         string `json:"http_addr,omitempty"` // e.g. "127.0.0.1", "unix:/tmp/app.sock"
	HttpPort         uint16 `json:"http_port,omitempty"` // e.g. 8080
	HttpTimeout      uint16 `json:"http_timeout,omitempty"`
	UrlBasePath      string `json:"url_base_path,omitempty"`
	CookieKeyLocale  string `json:"cookie_key_locale,omitempty"`
	CookieKeySession string `json:"cookie_key_session,omitempty"`
}
```


参数说明

| 字段 | 类型 | 默认值 | 必需 | 说明 |
|----|----|----|----|----|
| HttpAddr | string | 否 | 空 | 设置以 Unix domain socket 方式发布服务 |
| HttpPort | int | 否 | 8080 | 设置以 TCP 端口发布服务 |
| HttpTimeout | int | 否 |  30 | 设置 http 连接超时时间, 单位秒 |
| UrlBasePath | string | 否 | / | 设置 http 服务访问的URL根路径，默认为 / |
| CookieKeyLocale | string | 否 | lang | 当启用 i18n 时，httpsrv 会在cookie中以默认字段名 `lang` 设置语言包参数，这个值可自定义 cookie 保存的字段名 |
| CookieKeySession | string | 否 | access_token | 当启用 Session 时，httpsrv 会在cookie中以默认字段名 `access_token` 设置用户状态的 Session 值信息，这个值可自定义 cookie 保存的字段名 |

Config 是 [Service](service.md) 的一个内置项，通过 Service 引用, 如:

``` go
package main

import ( 
	"github.com/hooto/httpsrv"
)

func main() {
	// 通过全局 Service 实例引用 
	httpsrv.DefaultService.Config.HttpPort = 8080

	// 新建 Service 实例引用
	srv := httpsrv.NewService()
	srv.Config.HttpPort = 8081
}
```

## 扩展配置项

在 `type Config struct` 这个数据定义基础之上，扩展了部分动态接口用于扩展配置项

### 注册新的模版内置调用函数 (可选)

``` go
package main

import ( 
	"strings"
	"github.com/hooto/httpsrv"
)

func ExampleUpper(v string) string {
	return strings.ToUpper(v)
}

func main() {
	var conf httpsrv.Config
	// ...
	conf.RegisterTemplateFunc("upper", ExampleUpper)
}
```

> 注: 系统默认内置了常用视图模版(View Template)函数，具体可参考代码文件 [template-func.go](https://github.com/hooto/httpsrv/blob/master/template-func.go)



config.go
// Copyright 2015 Eryx <evorui at gmail dot com>, All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package httpsrv

const Version = "0.12.0"

type Config struct {
	// e.g. "127.0.0.1", "unix:/tmp/app.sock"
	HttpAddr string `json:"http_addr,omitempty" toml:"http_addr,omitempty"`

	// e.g. 8080
	HttpPort uint16 `json:"http_port,omitempty" toml:"http_port,omitempty"`

	HttpTimeout uint16 `json:"http_timeout,omitempty" toml:"http_timeout,omitempty"`

	UrlBasePath string `json:"url_base_path,omitempty" toml:"url_base_path,omitempty"`

	CookieKeyLocale  string `json:"cookie_key_locale,omitempty" toml:"cookie_key_locale,omitempty"`
	CookieKeySession string `json:"cookie_key_session,omitempty" toml:"cookie_key_session,omitempty"`

	CompressResponse bool `json:"compress_response,omitempty" toml:"compress_response,omitempty"`
}

var DefaultConfig = Config{

	HttpAddr: "0.0.0.0",
	HttpPort: 8080,

	HttpTimeout: 30, // 30 seconds

	CookieKeyLocale:  "lang",
	CookieKeySession: "access_token",
}

func (c *Config) RegisterTemplateFunc(name string, fn interface{}) {
	tplMut.Lock()
	defer tplMut.Unlock()
	TemplateFuncs[name] = fn
}

func (c *Config) I18n(file string) {
	i18nLoadMessages(file)
}
